### PR TITLE
DynamoDB Autoscaling - Use DependsOn to avoid CloudWatch API Rate Limits

### PR DESCRIPTION
**Pull request by ThatJoeMoore**
Cloudwatch has very low API Rate Limits. Turns out, it's pretty easy to
hit that limit when setting up lots of application autoscaling rules in
Cloudformation (see https://forum.serverless.com/t/autoscaling-for-amazon-dynamodb/2071). So, by adding a DependsOn to all of our scaling targets,
we can avoid hitting that rate limit by forcing Cloudformation to only
make one call at a time, rather than parallelizing all of them.

In order to make that happen, I've also restructured the autoscaling
template.  It should produce the same output as before, with added
'DependsOn' attributes, but with a lot less duplication in the
Handlebars template itself. Plus, the new structure makes it a LOT
easier to compute the 'DependsOn' values.

### CODE REVIEW

**Comment by dsw88** [comment]
The change looks great! I just had one thought about it: The index file is getting a little long at this point. What do you think about splitting the auto-scaling stuff out into its own file in this directory? It might help clarify the top-level deployer functionality to package that stuff away into its own module.

**Comment by ThatJoeMoore** [comment]
Yeah, I like that.  Updated PR incoming.

**Comment by ThatJoeMoore** [comment]
Alright, I've got the autoscaling config in its own (typescript!) file.

This commit also implicitly proposes a method for handling our
Typescript conversions, namely, creating a separate file (config-types.ts) containing only
the types used in configuring a given service. This allows us to split
some parts of a service implementation into separate files without
creating a circular dependency.

File: package.json [hunk 1]
```diff
@@ -52,8 +52,8 @@
     "run-sequence": "^2.2.0",
     "sinon": "^1.17.7",
     "sinon-chai": "^2.13.0",
+    "ts-node": "^3.3.0",
     "tslint": "^5.8.0",
-    "typescript": "^2.5.3",
-    "ts-node": "^3.3.0"
+    "typescript": "^2.5.3"
```
Comment (ThatJoeMoore) [inline package.json]: Looks like the latest version of NPM changed their alphabetizing rules...

File: src/datatypes/service-context.ts [hunk 2]
```diff
@@ -16,20 +16,21 @@
  */
 
 import { AccountConfig } from './account-config';
+import ServiceConfig from './service-config';
 
-export class ServiceContext {
+export class ServiceContext<Config extends ServiceConfig = any> {
```
Comment (ThatJoeMoore) [inline src/datatypes/service-context.ts]: This and the below changes let us optionally specify what type the params will be, but it leaves all current usages intact and unmodified. I really, really like having default values for types!

**Pull request merged**